### ISSUE TITLE
Object Labeler Functionality (updated feature request)

### ISSUE THREAD

**Issue by gerald-hartig** [nvaccess/nvda#17332]
Information transferred from old issue https://github.com/nvaccess/nvda/issues/2111

### Is your feature request related to a problem? Please describe.
Many applications contain unlabelled or poorly labelled controls (buttons, graphics, icons, etc.) that NVDA can only announce as "button" or "unknown." This creates significant accessibility barriers:

- Users must memorise the number of key presses to access specific controls
- Some workplace applications become unusable without proper labelling
- Users are forced to switch to other screen readers (like JAWS) that offer labelling capabilities
- This issue affects various applications including:
  - Enterprise software with icon-only interfaces
  - Printer/scanner utility applications
  - Applications with custom UI controls
  - Legacy applications with poor accessibility

### Describe the solution you'd like
Add a built-in object labelling system to NVDA that allows users to:

1. Assign custom labels to controls while navigating using object navigation
2. Save labels persistently between sessions
3. Access labelled controls consistently within the same application
4. Share labels with other users (optional community database)

Key requirements:
- Activate via keyboard shortcut (e.g., NVDA+G as originally proposed)
- Store labels in a human-readable format
- Support application-specific labelling
- Maintain NVDA's performance 
- Allow import/export of labels

### Describe alternatives you've considered
1. **Add-on Implementation**: Several add-ons have attempted to provide this functionality, but a core implementation would ensure:
   - Consistent behaviour
   - Better performance optimisation
   - Wider adoption
   - Proper maintenance

2. **Alternative Identification Methods**:
   - Using object properties instead of full paths where possible
   - Class-based identification before path-based
   - Caching mechanisms to improve performance

3. **Scope Limitations**:
   - Limiting labelling to specific application types
   - Excluding web content labelling (handled separately)
   - Focusing on truly inaccessible controls only

### Additional context
- Other screen readers (JAWS, TalkBack) successfully implement similar features
- This has been a long-requested feature since 2012 (original issue #2111)
- Performance concerns need to be carefully addressed:
  - Object path traversal can be expensive
  - Large numbers of labels could impact responsiveness
  - Need efficient storage and retrieval mechanisms
- Consider modern application architectures:
  - Single executable hosting multiple applications
  - Dynamic UI generation
  - Custom control frameworks
- Strong community demand across multiple language communities (Portuguese, Romanian, Chinese, Arabic, etc.)
- Could potentially integrate with the NVDA Developer Info tool to help identify objects consistently

**Comment by rkingett** [comment]
Along with this, a system to clear labels would be good to keep responsiveness at it's peak.

**Comment by haruncetinkaya306** [comment]
What I'm wondering is whether the labeling done by NVDA will be local. If the labeling is in English, this could be a bit of a problem for those speaking other languages.

**Comment by ABuffEr** [comment]
Hi,
it covers just a subset of the problem (labels present but not correctly associated to objects), but recently I created [this support module.](https://github.com/ABuffEr/labelAutofinderCore) It's not fully automatic, but simplify a lot the management of above case developing add-ons.
Hth.

**Comment by Gene703122** [comment]
I strongly support the proposal to have object label ability.  It has been clearly stated in the ticket the numerous advantages this would confer and that it would make NVDA much more usable in employment situations.  Also, as a separate project, I advocate that the ability to label links be incorporated.

**Comment by christodeklerk** [comment]
On an NVDA users email list there has just been a lengthy discussion about a labelling feature and it is clear that there is a real need for it. Although one person strongly opposed it, he could not provide any practical problems with implementing such a feature. His objections were rather philosophically based on the grounds that labelling screen items is the responsibility of software developers and not of screen readers. That is not going to happen soon, so why should screen reader users who really need the feature be deprived of it for no convincing practical reasons? I know blind persons who had to pay substantial amounts of money for Jaws, because it is able to label unlabeled screen elements in software they need to use productively in the work place. I appeal to NVDA developers not to allow this to drag on even longer, but to work on the implementation of this feature. The way it works in Jaws or used to work in Window-Eyes could be used as a model.

**Comment by josephsl** [comment]
Hi,

I am the one who have objected based on philosophical grounds. Later in the thread, I backed off and suggested that an add-on found by someone else could serve as a starting point to work on this feature.

At the moment I advise against suggesting a direct implementation in NVDA Core to reduce burden for NVDA contributors working on 64-bit transition. Rather, I suggest letting an add-on serve as a prototype for this feature similar to features such as screen curtain and sound splitter.

Thanks.

**Comment by britechguy** [comment]
[christodeklerk](https://github.com/christodeklerk)
> On an NVDA users email list there has just been a lengthy discussion

Those interested in that user feedback should read the topic:
[Is there a way to label things on a website like links and buttons?](https://nvda.groups.io/g/nvda/topic/is_there_a_way_to_label/116777640)

**Comment by seanbudd** [comment]
While add-ons are good for experimentation, this feature request has been triaged, and we encourage a direct pull request to NVDA for the feature. We don't think a PR for this would affect the 99% complete 64-bit migration, and we don't want to discourage community members from doing work they are interested in doing. We've been working with community members on a new magnifier feature and integrating MathCAT while doing the 64bit transition. I think too often add-ons are created and then left as add-ons. Their code ends up diverging from our standards, and then it becomes too much work for the add-on author to integrate it into NVDA core.

**Comment by christodeklerk** [comment]
I agree strongly with seanbudd's comments. It could become problematic if this feature would be available in an add-on rather than in the NVDA core. Add-ons often get abandoned. There are annual add-on breaking updates of NVDA and it often takes months for add-ons to be made compliant. Users who depend on this feature would then not be able to update NVDA until the add-on gets updated. I would urge developers to implement this feature as part of the NVDA core.

**Comment by id2025anc** [comment]
I strongly agree with the previous comments regarding the need for an object labelling feature in NVDA. At its core, this feature should allow the user to specify a custom label for a control such as a button and save these labels in an application-specific configuration profile. Beyond that, it would be great if the feature could utilize OCR or on-device or remote AI pattern and image analysis to suggest the most likely identification of any unlabeled controls, similar to what VoiceOver does in iOS and what the Viewpoint for Windows app does.

Such a feature can not only be critical in employment settings, but also would open doors in people's personal lives. For example, the Windows installer for NordVPN contains mostly unlabeled controls, and the installed Windows apps of NordVPN and ExpressVPN also contain some inaccessible controls.
